proposals/csharp-7.2/span-safety.md

```diff
@@ -290,6 +288,93 @@ We wish to ensure that no `ref` local variable, and no variable of `ref struct`
+
+```csharp
+void M1(ref Span<int> s1)
+{
+    Span<int> s2 = stackalloc int[1];
+    Swap(ref s1, ref s2);
+}
+
+void Swap(ref Span<int> x, ref int Span<int> y)
+{
+    // This will effectively assign the stackalloc to the s1 parameter and allow it
+    // to escape to the caller of M1
+    ref x = ref y; 
+}
+```
+
+The restriction on the receiver is necessary because while none of its contents are ref-safe-to-escape it can store the provided values. This means with mismatched lifetimes you could create a type safety hole in the following way:
+
+```csharp
+ref struct S
+{
+    public Span<int> Span;
+
+    public void Set(Span<int> span)
+    {
+        Span = span;
+    }
+}
+
+void Broken(ref S s)
+{
+    Span<int> span = stackalloc int[1];
+
+    // The result of a stackalloc is now stored in s.Span and escaped to the caller
+    // of Broken
+    s.Set(span); 
+}
+```
+
+### Struct This Escape
+When it comes to span safety rules the `this` value in an instance member is modeled as a parameter to the member. Now for a `struct` the type of `this` is actually `ref S` where in a `class` it's simply `S` (for members of a `class / struct` named S). 
+
+Yet `this` has different escaping rules than other `ref` parameters. Specifically it is not ref-safe-to-escape while other parameters are:
+
+```csharp
+ref struct S
+{ 
+    int Field;
+
+    // Illegal because this isn't safe to escape as ref
+    ref int Get() => ref Field;
+
+    // Legal
+    ref int GetParam(ref int p) => ref p;
+}
+```
+
+The reason for this restriction actually has little to do with `struct` member invocation. There are some rules that need to be worked out with respect to member invocation on `struct` members where the receiver is an rvalue. But that is very approachable. 
+
+The reason for this restriction is actually about interface invocation. Specifically it comes down to whether or not the following sample should or should not compile;
+
+```csharp
+interface I1
+{
+    ref int Get();
+}
+
+ref int Use<T>(T p)
+    where T : I1
+{
+    return ref p.Get();
+}
+```
+
+Consider the case where `T` is instantiated as a `struct`. If the `this` parameter is ref-safe-to-escape then the return of `p.Get` could point to the stack (specifically it could be a field inside of the instantiated type of `T`). That means the language could not allow this sample to compile as it could be returning a `ref` to a stack location. On the other hand if `this` is not ref-safe-to-escape then `p.Get` cannot refer to the stack and hence it's safe to return. 
+
+This is why the escapability of `this` in a `struct` is really all about interfaces. It can absolutely be made to work but it has a trade off. The design eventually came down in favor of making interfaces more flexible. 
+
+There is potential for us to relax this in the future though. 
```

Comment:
Interfaces allow the caller to be oblivious whether it works with classes or structs. That pushes the safety rules into the calee, where we know the kind of the containing type.

`ref struct` does not implement interfaces, so in theory could treat `this` as ordinary `ref` parameter. I am not 100% sure, but I think we decided against specialcasing `ref` structs vs ordinary structs.

This may not be as easy to relax as it seems. Primarily because you still need something for structs. You can only move where you enforce it - at the caller or at and callee, but such move would be compat breaking.

Reply:
I think there is actually a really easy way to relax this. Consider if we change the rules as follows:

> The `this` parameter in a `struct` instance member is ref-safe-to-escape if a) if the member does not implement an interface member and b) does not have any `out` or `ref` parameters. 
> Else it is ref-safe-to-escape to the method body only.

That would cover all the cases I'm aware of and be compatible with existing code.

Reply:
Then on the caller side you would need to consider a possibility of the method ref-returning `this`.  That could break things.

```cs
    public ref int M1()
    {
        Span<int> s = new int[]{42} ;

        // the following is currently allowed, 
        // but if indexer can do `return ref Length`, shouldn't this become an error?
        return ref s[0];
    }
```

Reply:
Hmm that is true as today we consider everything to be heap lifetime if it comes from a `struct` with no other returnable values. That would suggest we have to provide some manner of opt in for this to be allowed. That's unfortunate.